### PR TITLE
Update the usage in redis and replace redigo with go-redis. Use hashmap to cache the data.

### PR DESCRIPTION
**Pull request by hopkings2008**

### CODE REVIEW

**Comment by hopkings2008** [comment]
@thesues @cxt90730 please helper to review this PR.

**Comment by hopkings2008** [comment]
@thesues @cxt90730 i have tested the redis-sentinel mode and it works fine. i just tested the basic funcs by using go-redis in redis-sentinel env. i didn't test the failover by redis-sentinel.:)

File: meta/bucket.go [hunk 1]
```diff
@@ -61,10 +78,102 @@ func (m *Meta) GetBucketInfo(bucketName string) (Bucket, error) {
...
+		err = m.Client.UpdateUsages(bucketUsageCacheMap, nil)
+		if err != nil {
+			helper.Logger.Println(2, "failed to sync usages to database, err: ", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Meta) bucketUsageSync(bucketName string) error {
```
Comment (cannium) [inline meta/bucket.go]: Maybe add `usage` value to sync event message so no need to check from Redis again, meanwhile other improvements, like merging consecutive writes, become possible.

File: meta/client/tidbclient/bucket.go [hunk 2]
```diff
@@ -12,51 +12,62 @@ import (
 	"time"
 )
 
-func (t *TidbClient) GetBucket(bucketName string) (bucket Bucket, err error) {
+func (t *TidbClient) GetBucket(bucketName string) (bucket *Bucket, err error) {
 	var acl, cors, lc, policy, createTime string
-	sqltext := "select * from buckets where bucketname=?;"
+	var updateTime sql.NullString
+	sqltext := "select bucketname,acl,cors,lc,uid,policy,createtime,usages,versioning,file_counts,update_time from buckets where bucketname=?;"
+	tmp := &Bucket{}
```
[the remark this answers is not in the document]
Comment (hopkings2008) [reply meta/client/tidbclient/bucket.go]: tmp is for assign the value to bucket in the end. If there is some error in the middle, the return value bucket will be nil.

File: meta/bucket.go (hunk 1 above)
Comment (hopkings2008) [reply meta/bucket.go]: thanks @cannium for your good suggestions, i will encapsulate the bucket name and it's usage into the parameter and update it to db directly.
Consecutive writes is a good performance improvements suggestion. Can we make enhancement for it in the next step? :)

Comment (cannium) [reply meta/bucket.go]: Sure, do as your convenience.

Comment (hopkings2008) [reply meta/bucket.go]: @cannium, please help to review this PR again. I have amended the bucket sync logic and just transfer the bucket name and usage to update db directly. Thanks.

**Comment by thesues** [comment]
These patches requires "a never lose data redis". We could accept these patches into a new branch, and compare this solution to others.

**Pull request merged**